### PR TITLE
Now, we cache logprobs to disk in DPO. 

### PR DESCRIPTION
**Pull request by finbarrtimbers**
We use a hash based path, so that it will work across seeds. 

Sample run of single GPU DPO: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KEAE4VNFXT7DZNJ5Y2CY5RWJ?taskId=01KEAE4VNKQ3567BTJWM6RGZMF&jobId=01KEAE4VSJCXVVP04N5NDZNKSX).

---

> [!NOTE]
> Adds deterministic, hash-based on-disk caching of reference logprobs for DPO using a new TensorCache, wiring dataset_index through collators and updating the training loop; includes GPU/unit tests and a large DPO run script.
> 
> - **DPO training**:
>   - Add `compute_reference_logprobs_cache_hash` and `build_reference_logprobs_cache` to precompute and cache reference logprobs using a deterministic hash path.
>   - Introduce `args.reference_logprobs_cache_path` and compute cache key from model/dataset config (incl. packed/LoRA/QLoRA settings).
>   - Wire `dataset_index` into datasets and batches to index cached tensors; use cached logprobs in `dpo`/`dpo_norm`/`wpo` loss.
>   - Handle distributed aggregation of cached tensors with `torch.distributed.all_reduce` and persist/load cache from disk.
> - **Utilities**:
>   - Add `TensorCache` in `open_instruct/model_utils.py` with indexing, device move, and disk save/load.
> - **Collators**:
>   - Pass through `dataset_index` in `DataCollatorForSeq2SeqDPO` and `TensorDataCollatorWithFlatteningDPO`.
> - **Tests**:
>   - New GPU tests validating `TensorCache` GPU indexing/disk round-trip and collator `dataset_index`; tests for cache-hash determinism.
>   - Add unit tests for `TensorCache` CPU behaviors.
> - **Scripts/Docs**:
>   - Add `scripts/train/debug/large_dpo.sh` and run instructions in `.claude/commands/run-dpo-experiments.md`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 39f0e87077c8135dbf18cbac26a0aa35209479dc. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: open_instruct/dpo_tune_cache.py [hunk 1]
```diff
@@ -442,46 +449,81 @@ def build_deepspeed_config(
...
+            chosen_tensor[dataset_indices] = chosen_logps
+            rejected_tensor[dataset_indices] = rejected_logps
+
+    # Use MAX instead of SUM because Accelerate's distributed sampler can duplicate
+    # examples across ranks when even_batches=True. MAX works because duplicate
+    # indices compute identical logprobs, so MAX of identical values is correct.
+    # TODO(finbarr): Refactor to use HFDataLoader and explicitly shard the dataset to prevent duplicates.
+    if dist.is_initialized():
+        dist.all_reduce(chosen_tensor, op=dist.ReduceOp.MAX)
+        dist.all_reduce(rejected_tensor, op=dist.ReduceOp.MAX)
```
Comment (hamishivi) [inline open_instruct/dpo_tune_cache.py:520]: Isn't the dataloader sharded by accelerate? Sorry a bit confused by this.

File: open_instruct/dpo_tune_cache.py [hunk 2]
```diff
@@ -891,20 +956,20 @@ def load_model():
...
-        epoch_cached_reference_chosen_logps, epoch_cached_reference_rejected_logps = get_cache_ref_logprobs(
-            model,
-            train_dataloader,
-            accelerator,
-            average_log_prob,
-            last_checkpoint_path,
-            resume_step,
-            range(starting_epoch, args.num_train_epochs),
-            forward_fn,
+    if args.dpo_loss_type in ["dpo", "dpo_norm", "wpo"]:
```
Comment (hamishivi) [inline open_instruct/dpo_tune_cache.py:953]: oop, nice catch

Comment (finbarrtimbers) [reply open_instruct/dpo_tune_cache.py:953]: ty

File: open_instruct/dpo_tune_cache.py (hunk 1 above)
Comment (finbarrtimbers) [reply open_instruct/dpo_tune_cache.py:520]: The dataloader is, but not the dataset, so this should be fine! And because we use the `index`, we're good to go (we actually always were, even before my recent changes).

Comment (hamishivi) [reply open_instruct/dpo_tune_cache.py:520]: Yup, seems good!

**Review by hamishivi — APPROVED** [review]
